### ISSUE TITLE
Feature: add bindFailure

### ISSUE THREAD

**Issue by GregOnNet** [seangwright/typescript-functional-extensions#20]
**Is your feature request related to a problem? Please describe.**

- I was facing the following source code:

```ts
if (coindition_1) {
  return value_1;
}

if(condition_2) {
  return value_2
}

return null;
```

- I started refactoring it using typescript-functional Extensions.
- Therefore `condtion_1` & `condition_2` got extracted into own functions, returning a `Result`.

```ts
function condition_1(): Result {}
function condition_2(): Result {}
```

-  No, I wanted to compose these functions.
- If `condition_1()` succeeds, use its Result, otherwise execute `condition_2()`

I was not able to figure out how to achieve this with the built-in operators.
Maybe I am thinking too procedural, here?

**Describe the solution you'd like**
It would be nice having an operator, allowing to execute a function that returns a Result, if the predecessor-Result is a failure.

**Describe alternatives you've considered**
- Currently, I wrote a custom operator that mimics the described behaviour
- see [StackBlitz](https://stackblitz.com/edit/typescript-vga279?file=index.ts)
  - `tryRecoverIfFailure`

**Additional context**
I am wondering if it is beneficial to add an operator to the library's core.
Possibly there is already a better way, I am not able not see.

Thank you so much, Cheers
Gregor

**Comment by GregOnNet** [comment]
Hi,

today I was facing this use case again.
This time `ResultAsync` was involved, that's why I wrote an async version of `tryRecoverIfFailure`.

Not sure if this is the best possible implementation.
I post both implementations here, for documentation purpose.


### Result

```ts
export function tryRecoverIfFailure<TValue, TError>(projection: FunctionOfT<Result<TValue, TError>>): ResultOpFn<TValue, TError, TValue, TError> {
  return (result) => {
    return result.isFailure ? projection() : result;
  };
}
```

### ResultAsync

```ts
export function tryRecoverIfFailureAsync<TValue, TError>(
  projection: FunctionOfT<ResultAsync<TValue, TError>>,
): ResultAsyncOpFn<TValue, TError, TValue, TError> {
  return (resultAsync) => {
    return ResultAsync.from(resultAsync.isSuccess.then((isSuccess) => (isSuccess ? resultAsync.toPromise() : projection().toPromise())));
  };
}
```

<details>
<summary>Unit Tests</summary>
describe('typescript-functional-extensions', () => {
  describe('recoverIfFailureAsync', () => {
    describe('When the previous result fails', () => {
      it('takes the result from the second result', async () => {
        const resultAsyncRecoverySuccess = () => ResultAsync.success('✅');

        const resultAsync = ResultAsync.failure('💥').pipe(tryRecoverIfFailureAsync(() => resultAsyncRecoverySuccess()));
        const result = await resultAsync.toPromise();

        expect(result.getValueOrThrow()).toBe('✅');
      });
    });

    describe('When the previous result succeeds', () => {
      it('takes its result', async () => {
        const resultAsyncRecoveryIrrelevant = () => ResultAsync.failure('🤷‍♂️');

        const resultAsync = ResultAsync.success('✅').pipe(tryRecoverIfFailureAsync(() => resultAsyncRecoveryIrrelevant()));
        const result = await resultAsync.toPromise();

        expect(result.getValueOrThrow()).toBe('✅');
      });
    });

    describe('When both results fail', () => {
      it('takes the failure from the second result', async () => {
        const resultAsyncRecoveryFailure = () => ResultAsync.failure('💥💥');

        const resultAsync = ResultAsync.failure('💥').pipe(tryRecoverIfFailureAsync(() => resultAsyncRecoveryFailure()));
        const result = await resultAsync.toPromise();

        expect(result.getErrorOrThrow()).toBe('💥💥');
      });
    });
  });
});
</details>

**Comment by GregOnNet** [comment]
Short Update:

In my project, I renamed the operators to:
- bindFailure
- bindFailureAsync

This naming makes clear that the projection-function expects a ResultAsync.

**Comment by seangwright** [comment]
I'd take a PR for those 👍 - I like the name and I think they could be useful to more developers.

**Comment by GregOnNet** [comment]
That's great to here.

You can assign me to this issue. I will prepare the PR. 🥳

**Comment by GregOnNet** [comment]
@seangwright, PR #22 available

**Comment by seangwright** [comment]
Resolved with #22
